### DISCUSSION TITLE
Docusaurus v3 - Upgrade Support

### DISCUSSION THREAD

**Discussion by slorber** [facebook/docusaurus#9336] [General]
Docusaurus v3 is around the corner. 

Please use this discussion to ask questions so that we can help you upgrade.

## Other useful support links

- [Docusaurus v3 - Upgrade Support](https://github.com/facebook/docusaurus/discussions/9336)
- [Docusaurus v3 - Discord channel #migration-v2-to-v3](https://discord.com/channels/398180168688074762/1154771869094912090)
- [MDX v3 - Upgrade Support](https://github.com/facebook/docusaurus/discussions/9053)
- [MDX v3 - Remark/Rehype Plugins Support](https://github.com/facebook/docusaurus/discussions/9337)
- [MDX v3 - Discord channel #migration-mdx-v3](https://discord.com/channels/398180168688074762/1116724556976111616)

## Before asking for support

To ask for support, please:
- Check if someone else already asked a similar question
- Create a minimal repro using [docusaurus.new](https://docusaurus.new)
- Give us the fs path of the problematic Markdown file, and the URL pathname to access it
- Spend some time explaining your problem clearly in a structured and concise way
- Paste full, clear error messages and provide screenshots if appropriate

It's not scalable to provide done-for-you support for all Docusaurus users looking to upgrade. Any support request not respecting those constraints will possibly be ignored. 

**If you want help, you need to invest your own time so that helping you is easy.**

Alternatively, you can look for a paid [Docusaurus Service Provider](https://github.com/facebook/docusaurus/discussions/9281) to help you.

**Comment by johnnyreilly** [comment]
Reported an issue here: https://github.com/facebook/docusaurus/issues/9379

**EDIT**: issue about React hydration errors being reported to the console

**Comment by johnnyreilly** [comment]
I've been testing out the Docusaurus 3 beta which migrates to MDX 2. I have a couple of rehype plugins that improve core webvitals and serve images with cloudinary. This post walks through the migration for each plugin:

https://johnnyreilly.com/docusaurus-3-how-to-migrate-rehype-plugins

**Reply by slorber** [reply]
Thanks

Worth posting this in https://github.com/facebook/docusaurus/discussions/9337 instead ;) 
I suppose most Docusaurus users don't create Rehype plugins. Will link to it in the support intro

**Comment by timolagus** [comment]
I just upgraded from Docusaurus `2.4.1` to `3.0.0`. Everything went splendidly with the exception of one issue:

**Custom markup no longer renders in the sidebar.**

In the `sidebars.js` file, I have the following variable for a separator:

```js
const homeSeparator = {
  type: 'html',
  value: '<hr />',
  defaultStyle: true,
};
```

I use the variable to separate a section home page from the rest of the pages in that section, for example:

```js
const sidebarGettingStarted = [
  'getting-started/getting-started',
  homeSeparator,
  'getting-started/about-this-documentation',
  {
  // ...
```

In `2.4.1`, this renders a separator in the sidebar as expected:

![image](https://github.com/facebook/docusaurus/assets/143693948/bc7b4a3c-62f0-468f-98a4-0205a1a02f16)

The build source code contains the following `<li>` element for the separator:

```html
<li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menuHtmlItem_M9Kj menu__list-item">
    <hr>
</li>
```

In `3.0.0`, the separator is missing:

![image](https://github.com/facebook/docusaurus/assets/143693948/213e9345-8919-4761-87fc-8250347ea38e)

In the source code, there's no `<li>` element or anything for the separator.

I get no error messages about this during building (`docusaurus build` or `docusaurus start`). It seems like the build simply ignores the `html` item type in the sidebar.

I can reproduce the issue with your [skeleton Docusaurus website](https://docusaurus.io/docs/installation#scaffold-project-website) by adding my separator code to the default `sidebars.js` file, for example:

```
/** @type {import('@docusaurus/plugin-content-docs').SidebarsConfig} */
const sidebars = {
  // By default, Docusaurus generates a sidebar from the docs folder structure
  // tutorialSidebar: [{type: 'autogenerated', dirName: '.'}],

  // But you can create a sidebar manually
  tutorialSidebar: [
    'intro',
    // 'hello',
    {
      type: 'html',
      value: '<hr />',
      defaultStyle: true,
    },
    {
      type: 'category',
      label: 'Tutorial',
      items: ['tutorial-basics/create-a-document'],
    },
  ],
};

export default sidebars;

```

When I build, I get no separator on the page:

![image](https://github.com/facebook/docusaurus/assets/143693948/03a7e10f-7320-4c32-abd1-dca1b94d4e48)

**Can you help me with this?**

I have a sinking feeling that I'm missing something obvious here, but I just don't know what that is. I did not see any mention in the release notes or other documentation of sidebar custom markup getting deprecated, for example.

**Reply by slorber** [reply]
About Docs HTML sidebar items disappearing.
 
This is a bug reported here: https://github.com/facebook/docusaurus/issues/9483
PR: https://github.com/facebook/docusaurus/pull/9531

Should be fixed in canary or upcoming v3.0.1

**Comment by kammo52** [comment]
**ERROR
Cannot read properties of undefined (reading 'tokenizePlaceholders')**
TypeError: Cannot read properties of undefined (reading 'tokenizePlaceholders')
    at eval (webpack-internal:///./node_modules/prismjs/components/prism-php.js:339:40)
    at Object.run (webpack-internal:///./node_modules/prism-react-renderer/dist/index.mjs:525:15)
    at eval (webpack-internal:///./node_modules/prism-react-renderer/dist/index.mjs:2962:28)
    at mountMemo (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:17225:19)
    at Object.useMemo (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:17670:16)
    at useMemo (webpack-internal:///./node_modules/react/cjs/react.development.js:1651:21)
    at useTokenize (webpack-internal:///./node_modules/prism-react-renderer/dist/index.mjs:2951:56)
    at Highlight (webpack-internal:///./node_modules/prism-react-renderer/dist/index.mjs:2980:18)
    at renderWithHooks (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:16305:18)
    at mountIndeterminateComponent (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:20069:13)

**My package.json**
{
  "name": "docusaurus",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "docusaurus": "docusaurus",
    "start": "docusaurus start",
    "build": "docusaurus build",
    "swizzle": "docusaurus swizzle",
    "deploy": "docusaurus deploy",
    "clear": "docusaurus clear",
    "serve": "docusaurus serve",
    "write-translations": "docusaurus write-translations",
    "write-heading-ids": "docusaurus write-heading-ids"
  },
  "dependencies": {
    "@docusaurus/core": "^3.0.0",
    "@docusaurus/plugin-google-gtag": "^3.0.0",
    "@docusaurus/plugin-google-tag-manager": "^3.0.0",
    "@docusaurus/plugin-ideal-image": "^3.0.0-rc.1",
    "@docusaurus/plugin-sitemap": "file:latest││@docusaurus/preset-classic@latest",
    "@docusaurus/preset-classic": "^3.0.0",
    "@docusaurus/theme-mermaid": "^3.0.0",
    "@mdx-js/react": "^3.0.0",
    "classnames": "^2.3.2",
    "clsx": "^1.2.1",
    "docusaurus-lunr-search": "^3.3.0",
    "hast-util-is-element": "^1.1.0",
    "prism-react-renderer": "^2.2.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "rehype-katex": "^7.0.0",
    "remark-math": "^6.0.0"
  },
  "devDependencies": {
    "@docusaurus/module-type-aliases": "^3.0.0",
    "@docusaurus/types": "3.0.0"
  },
  "browserslist": {
    "production": [
      ">0.5%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "engines": {
    "node": ">=18.0"
  }
}

**I only encounter this issue when I open the blogs.**

**Reply by slorber** [reply]
@kammo52 for this kind of report, it is impossible to help you solve your problem without a runnable repro using docusaurus.new

**Comment by mderriey** [comment]
Hi there 👋

We upgraded our docs site to Docusaurus v3 and had very few issues thanks to the grat upgrade guide you put together, big kudos!

We're facing a minor issue: admonitions don't render in block quotes in MDX files.
We don't know whether it's due to the MDX upgrade or a change in Docusaurus.
What we do know is this was working in Docusaurus v2.4.3.

Repro:

- StackBlitz: <https://stackblitz.com/edit/github-qijbs8?file=docs%2Ftutorial-basics%2Fmarkdown-features.mdx>
- When the site is up, open the menu in the top-left corner, expand "Tutorial - Basics", then "Markdown features"
- Go to the "Admonitions" section
- Observe two new admonitions, and the one within the blockquote doesn't render

MDX markup:

```mdx
:::caution There be dragons

This is the admonition content

:::

> :::caution There be dragons
>
> This is the admonition content
>
> :::
```

Rendered:

![image](https://github.com/facebook/docusaurus/assets/6102639/0a26bb5a-bc65-49a3-a97f-475032de1a78)

_We know the `caution` admonition has been deprecated in v3 and should be replaced; this code is from a generation tool that we don't own, we'll open a separate issue with it to get it replaced by `warning`. Replacing `caution` with `warning` has the same result._

**Reply by slorber** [reply]
Thanks!

This bug has already been reported/fixed and will be released in v3.0.1 soon

https://github.com/facebook/docusaurus/pull/9535

**Reply by mderriey** [reply]
Thanks for coming back to me so quickly @slorber.

I took a look at the issue, and it didn't seem to quite fit my use case.
It's not just spaces and tabs before the admonition, but `>` characters for a blockquote.

I installed the latest canary (0.0.0-5748 at the time of writing) as a test and the problem persists.

**Reply by slorber** [reply]
Thanks, I'll add a fix for that quotes case

However you'd rather adopt the markdown directlve syntax directly if you only have a few cases like this, cf https://github.com/facebook/docusaurus/issues/9507#issuecomment-1806875146

This should work:

```mdx
> :::caution[There be dragons]
>
> This is the admonition content
>
> :::
```

**Reply by slorber** [reply]
Fixed in https://github.com/facebook/docusaurus/pull/9570

**Reply by mderriey** [reply]
Oh right, I hadn't clicked that it was a syntax issue, thanks.

I'll file an issue on the Docusaurus plugin we're using to see if we can move off the `caution` admonition and use the Markdown directive syntax.

Thanks again!

**Comment by homotechsual** [comment]
I'm in a bit of a sticky spot - a critical plugin I rely on breaks on v3 but with no errors so I'm at a loss to work out what needs to change to get it working on v3 sites which is holding back some key sites I'd love to get onto v3.

The plugin is: https://github.com/christian-bromann/docusaurus-theme-github-codeblock

Any insights would be gratefully received :-)

**Reply by slorber** [reply]
Answered in https://github.com/christian-bromann/docusaurus-theme-github-codeblock/issues/49

**Comment by RRQM** [comment]
Thank you for the long-term updates and support from Docusaurus. My open-source project TouchSocket has been deployed using the Docusaurus website template. It is very beautiful and stable.
When I learned that Docusaurus had released v3, I was very happy and couldn't wait to try a new version. But I encountered a problem upgrading. I have used the method shown in the upgrade document, but the problem has not been resolved yet.
So I came here to request help. I hope you can help me upgrade the website version to v3. I will be very grateful to Docusaurus.

github

https://github.com/RRQM/TouchSocket

doc

https://github.com/RRQM/TouchSocket/tree/master/handbook

**Reply by slorber** [reply]
Hey

You don't describe the problem? 😅

**Reply by RRQM** [reply]
I've tried it once, and I've got a lot of errors. It seems impossible to solve.

**Reply by RRQM** [reply]
If you can, can you please help me PR a branch?

**Reply by slorber** [reply]
@RRQM what you are asking here is me doing all the work for you. If you want that, you have to pay 😅 

I will only help for free if you show me your intention to try your best to solve problems on your own, and that you did your homework of reading carefully the migration guide and giving it a fair try. 

If you encounter a specific error, please describe that error and show us the inputs (Markdown file content, git branch, url to look at) that lead to that error. 

Only then we'll be able to help for free, one error at a time.

**Comment by homotechsual** [comment]
No problem!

________________________________________
